[FLINK-16510] Allow configuring shutdown behavior to avoid JVM freeze

mxm opened the pull request:
## What is the purpose of the change

This adds the configuration option `cluster.processes.halt-on-fatal-error`
which defaults to `false`. If set to `true`, a custom SecurityManager will be
installed on top of the existing SecurityManager to exit forcefully via
`Runtime#halt`.

Exiting that way may be necessary due to the Java 8 JVM freezing during a
graceful shutdown when using the G1 garbage collector.

## Brief change log

- If `cluster.processes.halt-on-fatal-error: true` is set in the config,
the shutdown will be performed via `Runtime#halt` instead of `System#exit`.

## Verifying this change

This change added tests and can be verified as follows:
  - Addded unit tests
  - Added integration test

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs / auto-generated configuration

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:
```diff
@@ -79,6 +79,16 @@
+	 */
+	@Documentation.Section(Documentation.Sections.ALL_TASK_MANAGER)
+	public static final ConfigOption<Boolean> GRACEFUL_SHUTDOWN_ON_ERROR =
+		key("taskmanager.graceful-shutdown-on-error")
```
I think it would be nice to make this behaviour configurable for all Flink processes.

mxm replied:
The scope is currently all task manager processes but it makes sense to extend it to all entry points.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java:
```diff
@@ -470,4 +479,30 @@ static String getTaskManagerResourceID(Configuration config, String rpcAddress,
+		try {
+			System.setSecurityManager(forcefulShutdownManager);
+		} catch (Exception e) {
+			throw new IllegalConfigurationException(
+				String.format("Could not register forceful shutdown handler for configuration '%s'. Either allow setting a SecurityManager or set the configuration to its default: '%s'",
+					GRACEFUL_SHUTDOWN_ON_ERROR.key(),
+					GRACEFUL_SHUTDOWN_ON_ERROR.defaultValue()),
+				e);
```
What if some corporate policies does not allow to install a custom `SecurityManager`?

mxm replied:
That is indeed a limitation of this approach. The default behavior doesn't require installing a security manager but to use the forceful shutdown feature, we need permission to add a security manager. The situation is similar for `Thread.setDefaultUncaughtExceptionHandler` because it also needs permission from the `SecurityManager`. Arguably, maybe that's a less restrictive permission.

tweise replied:
I think that is an acceptable limitation since it is an optional feature. It would be good to document it as part of the configuration option.

mxm commented:
Good questions @tillrohrmann. I was looking into setting a default exception handler but then realized we have exception handling in numerous places and it was not exactly trivial to figure out if the default handler would end up being called. Also, you mentioned an application can set a default exception handler too. 

In the end, the SecurityManager approach seemed to be the most elegant. It has the drawback of needing the permission to install a SecurityManager. Similarly, `Thread#setDefaultUncaughtExceptionHandler` also needs permission from the SecurityManager, if present. 

There is a bit of magic involved here, but not by default. Overall, I think it is reasonable to allow converting exit calls if explicitly switched on. After all, the configuration option is a workaround for a JVM bug which means that a bit of magic might be appropriate.

We can expand the configuration option to be a global option and not only affect the task managers.

tillrohrmann commented:
Fair points. If it should ever become a problem that we set a custom `SecurityManager` we might have to revisit this problem.

+1 for making the mechanism available for all Flink processes.

One small comment for the place where to instantiate the `SecurityManager`: At the moment it happens in the `TaskManagerRunner.startTaskManager` method. This method will be called multiple times by the `MiniCluster`, though. Hence, I would suggest to move the instantiation of the `SecurityManager` more to the entrypoints (`ClusterEntrypoint`, `MiniCluster`, `TaskManagerRunner.runTaskManagerSecurely` (update `YarnTaskExecutorRunner` to call into the same method)).

mxm commented:
PR updated to address the comments. 

- Added more reasoning to why this feature might be used
- Moved the initialization code
- I've skipped `MiniCluster` because I'm not sure this feature should be applied in a MiniCluster context. 
- Refactored startup for YarnTaskExecutorRunner
- I've renamed the configuration option name (defaults to false now)
- Added tests to check that the feature is disabled by default

mxm commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnTaskExecutorRunner.java:
```diff
@@ -90,17 +85,9 @@ private static void runTaskManagerSecurely(String[] args) {
-				TaskManagerRunner.runTaskManager(configuration, pluginManager);
-				return null;
-			});
+			TaskManagerRunner.runTaskManagerSecurely(configuration);
```
Good point!

mxm commented:
Updated and rebased.

zentol commented on flink-core/src/main/java/org/apache/flink/configuration/ClusterOptions.java:
```diff
@@ -78,4 +79,13 @@
+
+	@Documentation.Section(Documentation.Sections.EXPERT_CLUSTER)
+	public static final ConfigOption<Boolean> HALT_ON_FATAL_ERROR =
+		key("cluster.processes.halt-on-fatal-error")
```
You will need to regenerate the documentation.

mxm replied:
I'll update the docs once my computer has finished downloading the internet ;)

mxm replied:
Currently at:
```
[ERROR] npm ERR! code E404
[ERROR] npm ERR! 404 Not Found: @yarnpkg/lockfile@1.1.0
```

zentol commented on flink-annotations/src/main/java/org/apache/flink/annotation/docs/Documentation.java:
```diff
@@ -91,6 +91,7 @@
 		public static final String EXPERT_ZOOKEEPER_HIGH_AVAILABILITY = "expert_high_availability_zk";
 		public static final String EXPERT_SECURITY_SSL = "expert_security_ssl";
 		public static final String EXPERT_ROCKSDB = "expert_rocksdb";
+		public static final String EXPERT_CLUSTER = "expert_cluster";
```
You need to include this section into the `config.md` for it to show up.

mxm replied:
Thanks. Will do!

zentol commented on flink-core/src/main/java/org/apache/flink/configuration/ClusterOptions.java:
```diff
@@ -78,4 +79,13 @@
+		key("cluster.processes.halt-on-fatal-error")
+			.booleanType()
+			.defaultValue(false)
+			.withDescription("Whether processes should halt on fatal errors instead of performing a graceful shutdown. " +
+				"In some environments (e.g. Java 8 with the G1 garbage collector), a regular graceful shutdown can lead " +
+				"to a JVM deadlock. See: https://issues.apache.org/jira/browse/FLINK-16510");
```
```suggestion
			.withDescription(Description.builder().text(
				"Whether processes should halt on fatal errors instead of performing a graceful shutdown. " +
					"In some environments (e.g. Java 8 with the G1 garbage collector), a regular graceful shutdown can lead " +
					"to a JVM deadlock. See %s for details.",
				link("https://issues.apache.org/jira/browse/FLINK-16510", "FLINK-16510")
			.build());
```

mxm commented:
I'm unable to build the documentation. Attaching a log here in case there is something obvious I'm missing: [08-06T11_04_59_915Z-debug.log](https://github.com/apache/flink/files/5034433/2020-08-06T11_04_59_915Z-debug.log)

I used the following command:

```
mvn package -Dgenerate-config-docs -pl flink-docs -am -nsu -DskipTests -Dcheckstyle.skip
```

which currently yields

```
[ERROR] npm ERR! code E404
[ERROR] npm ERR! 404 Not Found: @antv/g2@3.4.10
[ERROR]
[ERROR] npm ERR! A complete log of this run can be found in:
[ERROR] npm ERR!     /Users/max/.npm/_logs/2020-08-06T11_04_59_915Z-debug.log
```

mxm commented:
Actually, it might not be me: https://status.npmjs.org/

![image](https://user-images.githubusercontent.com/837221/89525465-87433500-d7e6-11ea-905a-23d8bd65bc3d.png)

tillrohrmann commented:
Let me try whether I can build the documentation.

mxm commented:
Awesome, thanks for your review and for generating the docs!

zentol commented:
@mxm For the future, if you are running into issues building the WebUI, then you can skip it using `-Pskip-webui-build`.
